### PR TITLE
FOSSA scanning enabled for the repo

### LINKED ISSUE #854: Fix FOSSA for all repositories
UPDATE: 2020-09-01 main repo FOSSA check was fixed in #2347, but we need to add them to client libs as well.

* [ ] Java https://github.com/jaegertracing/jaeger-client-java
* [ ] Go https://github.com/jaegertracing/jaeger-client-go
* [ ] Go lib https://github.com/jaegertracing/jaeger-lib
* [ ] Python https://github.com/jaegertracing/jaeger-client-python
* [ ] Node.js https://github.com/jaegertracing/jaeger-client-node
* [ ] C/C++ (https://github.com/jaegertracing/cpp-client)
* [ ] C# / dotNET (https://github.com/jaegertracing/jaeger-client-csharp)

We used to have FOSSA checks run on many repositories, but now it only runs on 5 of them (excluding the main one), and we cannot add more.

According to Kevin Wang from FOSSA:

> I’ve noticed that Jaeger is enrolled in Automated Builds, which is not the ideal integration method. 
> 
> In this model, fossa has to “guess” your dependencies by running every build path — its convenient since you can import code through the UI, but not scalable if you want to make continuous scans. You will likely experience performance issues and false positives if you run this on a large project per commit.
> 
> I suggest we integrate Jaeger through Provided Builds, where you deploy our build client (https://github.com/fossas/fossa-cli) into your CI to report dependencies back to fossa. This model should enable extremely fast and high scale builds/scans from within your CI.
> 
> See “from a local dev machine” in this article below:
> 
> https://docs.fossa.io/docs/importing-a-project

### PR DESCRIPTION
**Pull request by idvoretskyi**
Fixes #854 

Signed-off-by: Ihor Dvoretskyi <ihor@linux.com>

### CODE REVIEW

File: .github/workflows/fossa.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: "^1.13.1"
+      - run: go version
+      # Runs a set of commands to initialize and analyze with FOSSA
+      - name: run FOSSA analysis
+        env:
+          FOSSA_API_KEY: '304657e2357ba57b416b94e6b119131b'
```
Comment (yurishkuro) [inline .github/workflows/fossa.yml:20]: What is the origin of this key? Please add a comment describing it.

File: .github/workflows/fossa.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    branches: [master]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: "^1.13.1"
```
Comment (yurishkuro) [inline .github/workflows/fossa.yml:15]: will it take `1.14.x`?

File: .github/workflows/fossa.yml (hunk 1 above)
Comment (idvoretskyi) [reply .github/workflows/fossa.yml:20]: @yurishkuro it's a FOSSA's push-only API token - https://docs.fossa.com/docs/api-reference#push-only-api-token; which is safe to be exposed publicly.

File: .github/workflows/fossa.yml (hunk 2 above)
Comment (idvoretskyi) [reply .github/workflows/fossa.yml:15]: @yurishkuro let's try :)

**Comment by idvoretskyi** [comment]
> Seems that it also timed out. When it says "still being analyzed on fossa.com", is the analysis done locally in CI or remotely via FOSSA?

@yurishkuro remotely.

**Comment by idvoretskyi** [comment]
@yurishkuro great, so it failed now due to the licensing issues found by FOSSA, not because of the wrong configuration.

**Comment by yurishkuro** [comment]
@idvoretskyi a couple questions:
  1. is there a way to tweak the GH action so that it posts the link to the report? For example, when you click on Details of the Codecov checks it takes you to their code coverage report, but FOSAA details take you to GH action log and you have to hunt for the URL of the report. Ideally it would be linked directly from Details, or even better posted as a summary as comment, like https://github.com/jaegertracing/jaeger/pull/2347#issuecomment-659639737
  2. The link from the logs  shows NotFound, why is that? If we run this scan on all PRs, then anyone should be able to view the reports.

https://app.fossa.com/projects/custom+162%2Fgithub.com%2Fjaegertracing%2Fjaeger/refs/branch/HEAD/da0020c14818c479a6272614e192674a7ad48993

image

**Comment by idvoretskyi** [comment]
@yurishkuro 

1. I'll check this and get back to you;
2. It's the default setting, let me take a look at how can it be updated.

File: .github/workflows/fossa.yml (hunk 1 above)
Comment (yurishkuro) [reply .github/workflows/fossa.yml:21]: but where does it come from? Is this ID specific to Jaeger project?

Comment (idvoretskyi) [reply .github/workflows/fossa.yml:21]: Yes, it was created for Jaeger only.

**Comment by idvoretskyi** [comment]
@yurishkuro I've updated the FOSSA configuration here, and PR won't be blocked by the code scanning issues - they will be handled separately and should not block the PRs.

Regarding the report access - access is limited to the FOSSA UI users only, I'll go ahead and invite you (and other project maintainers) so you could have access to them.

**Comment by idvoretskyi** [comment]
@yurishkuro anything else is missing to merge the PR?

**Comment by yurishkuro** [comment]
I am still getting Not Found when I go to the report link shown in the logs: https://app.fossa.com/projects/custom+162%2Fgithub.com%2Fjaegertracing%2Fjaeger/refs/branch/HEAD/80a27b16e68b355ee80308f8f36e6f3ff1391541 (even after I log in to Fossa).

**Comment by idvoretskyi** [comment]
@yurishkuro I don't think you've joined the CNCF FOSSA org, this is an issue.

I'll DM you with further details.

**Comment by idvoretskyi** [comment]
@yurishkuro anything else is needed to get this merged?

**Comment by yurishkuro** [comment]
@idvoretskyi I am confused about how this GH action is supposed to work. Before I rebased this just now, the Fossa step was showing green while the [report][1] that it was pointing to was showing 2 license issues. Shouldn't the action be shown as failed in that case?

[1]: https://app.fossa.com/projects/custom+162%2Fgithub.com%2Fjaegertracing%2Fjaeger/refs/branch/HEAD/80a27b16e68b355ee80308f8f36e6f3ff1391541

**Comment by idvoretskyi** [comment]
@yurishkuro no, we want this action to just collect reports for now. There's no need to block PRs by FOSSA reports.

**Comment by yurishkuro** [comment]
Whether to block PR or not is a setting in our repo, we can make FOSSA checks non-mandatory. But for maintainers it would be good to see the actual outcome of the check when approving PR. Right now doing that is pretty involved - not only the check comes back green, but also the report itself is quite hard to get to, you have to fish out the link from one of the step logs in the GH action. As I mentioned previously, it would be much better if
   1. the check would go red if there are license violations
   1. the link next to the check would go to a report directly, OR there would be a comment posted back into PR with a brief summary, similar to code coverage https://github.com/jaegertracing/jaeger/pull/2347#issuecomment-659639737

Having said that, an always green check is not going to hurt anything, so I am not opposed to merging this.

@objectiser @jpkrohling @pavolloffay any thoughts?

**Comment by objectiser** [comment]
+1 sounds ok to merge, but create an issue to make sure the actions you have suggested are investigated.

**Comment by idvoretskyi** [comment]
Umbrella issue (for the record) - https://github.com/cncf/foundation/issues/109

**Comment by jpkrohling** [comment]
Can we set a target date for us to decide to keep or remove it? Having it always green even on license violations might mean that we just won't care about it at all in the future... Suggestion: by v1.21.0, we decide if we want to keep it.

**Comment by yurishkuro** [comment]
Technically, we've been in this situation for several years already - the scans were running somewhere in the background, and the badge in the README kind-of reflects them, but we didn't pay attention (because it was always so noisy). The ticket #2432 lists the additional improvements.

**Pull request merged**